build-system/tasks/compile.js

```diff
@@ -353,98 +346,103 @@ function compile(entryModuleFilenames, outputDir, outputFilename, options) {
+      // args such as FORTESTING to our runner.
+      jscomp_off: ['unknownDefines'],
+      define,
+      hide_warnings_for: hideWarningsFor,
     };
 
     // For now do type check separately
-    if (argv.typecheck_only || checkTypes) {
+    if (options.checkTypes || options.typeCheckOnly) {
       // Don't modify compilation_level to a lower level since
       // it won't do strict type checking if its whitespace only.
-      compilerOptions.compilerFlags.define.push('TYPECHECK_ONLY=true');
-      compilerOptions.compilerFlags.jscomp_error.push(
+      compilerOptions.define.push('TYPECHECK_ONLY=true');
+      compilerOptions.jscomp_error.push(
           'checkTypes',
           'accessControls',
           'const',
           'constantProperty',
           'globalThis');
-      compilerOptions.compilerFlags.conformance_configs =
+      compilerOptions.conformance_configs =
           'build-system/conformance-config.textproto';
     }
 
-    if (compilerOptions.compilerFlags.define.length == 0) {
-      delete compilerOptions.compilerFlags.define;
+    if (compilerOptions.define.length == 0) {
+      delete compilerOptions.define;
     }
 
-    let stream = gulp.src(srcs)
-        .pipe(closureCompiler(compilerOptions))
-        .on('error', function(err) {
-          const {message} = err;
-          console./*OK*/error(colors.red(
-              'Compilation failed for ' + outputFilename + ':\n') +
-              formatClosureCompilerError(message));
-          process.exit(1);
-        });
-    // If we're only doing type checking, no need to output the files.
-    if (!argv.typecheck_only && !options.typeCheckOnly) {
-      stream = stream
-          .pipe(rename(outputFilename))
-          .pipe(replace(/\$internalRuntimeVersion\$/g, internalRuntimeVersion, 'runtime-version'))
+    // Required in order to override the JAR used by closure compiler
+    const platformOptions = {
+      platform: ['java'],
+    };
+
+    // Override to local closure compiler JAR
+    closureCompiler.compiler.JAR_PATH =
+        require.resolve('../runner/dist/runner.jar');
+
+    const gulpClosureCompiler = closureCompiler.gulp({
+      extraArguments: ['-XX:-TieredCompilation'],
+    });
+
+    const handleCompilerError = function(err) {
+      const {message} = err;
+      console./*OK*/error(colors.red(
+          'Compilation failed for ' + outputFilename + ':\n') +
+          formatClosureCompilerError(message));
+      process.exit(1);
+    };
+
+    if (options.typeCheckOnly) {
+      return gulp.src(srcs, {base: '.'})
+          .pipe(gulpClosureCompiler(compilerOptions, platformOptions))
+          .on('error', handleCompilerError)
+          .pipe(nop())
+          .on('end', resolve);
+    } else {
+      return gulp.src(srcs, {base: '.'})
+          .pipe(sourcemaps.init({loadMaps: true}))
+          .pipe(replace(
+              /\$internalRuntimeVersion\$/g, internalRuntimeVersion,
+              'runtime-version'))
```

Comment:
Bug 2: The calls to `replace` and `shortenLicense` must come _after_ closure compiler has run. The way we're doing things now, the text replacement is being done `O(N^2)` times on every input file, instead of `O(N)` times on every output file.

Reply:
This will break sourcemaps.

Reply:
@jridgewell Running `replace` and `shortenLicense` before closure compiler increases compile time by 60%. Can you talk about how sourcemaps breaks? Maybe there's a way to have our cake and eat it too?

Reply:
Figured it out offline. We can avoid the 60% slowdown by doing what I say in the top comment in this thread, and avoid the sourcemaps breakage by merging #21832.